components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -2377,6 +2404,142 @@ private static JWT encryptWithRSA(JWTClaimsSet jwtClaimsSet, JWEAlgorithm encryp
+     * @param clientId            ID of the client
+     * @return encrypted JWT token
+     * @throws IdentityOAuth2Exception
+     */
+    private static JWT encryptWithRSA(String signedJwt, JWEAlgorithm encryptionAlgorithm,
+                                      EncryptionMethod encryptionMethod, String spTenantDomain, String clientId)
+            throws IdentityOAuth2Exception {
+
+        try {
+            if (StringUtils.isBlank(spTenantDomain)) {
+                spTenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+                if (log.isDebugEnabled()) {
+                    log.debug("Assigned super tenant domain as signing domain when encrypting id token for " +
+                            "client_id: " + clientId);
+                }
+            }
+            String jwksUri = getSPJwksUrl(clientId, spTenantDomain);
+            Certificate publicCert;
+            JWK jwk;
+            Key publicKey;
+            String thumbPrint;
+
+            if (StringUtils.isBlank(jwksUri)) {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("Jwks uri is not configured for the service provider associated with " +
+                            "client_id: %s. Checking for x509 certificate", clientId));
+                }
+                publicCert = getX509CertOfOAuthApp(clientId, spTenantDomain);
+                thumbPrint = getThumbPrint(publicCert);
+                publicKey = publicCert.getPublicKey();
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("Fetching public keys for the client %s from jwks uri %s",
+                            clientId, jwksUri));
+                }
+                jwk = getValidJWKFromJWKS(jwksUri);
+                RSAKey rsaKey = RSAKey.parse(jwk.toJSONString());
+                publicKey = rsaKey.toRSAPublicKey();
+
+                if (jwk.getKeyID() != null) {
+                    thumbPrint = jwk.getKeyID();
+                } else {
+                    publicCert = getPublicCertFromJWK(jwk);
+                    thumbPrint = getJwkThumbPrint(publicCert);
+                }
+            }
+
+            return encryptWithPublicKey(publicKey, signedJwt, encryptionAlgorithm, encryptionMethod,
+                    spTenantDomain, clientId, thumbPrint);
+
+        } catch (JOSEException | ParseException e) {
+            throw new IdentityOAuth2Exception("Error occurred while encrypting JWT for the client_id: " + clientId
+                    + " with the tenant domain: " + spTenantDomain, e);
+        }
+    }
+
+
+    /**
+     * Get valid jwk from JWKS url when kid and JWKS Uri is given.
+     *
+     * @param jwksUri - JWKS Uri
+     * @return - valid JWK from the jwks url
+     * @throws IdentityOAuth2Exception - IdentityOAuth2Exception
+     */
+    private static JWK getValidJWKFromJWKS(String jwksUri) throws IdentityOAuth2Exception {
```

Comment:
Need a better name to reflect the purpose?

Is this a generic method? or a specific one related to encryption?

Reply:
changed the name to getEncryptionJWKFromJWKS() as this method is specific for fetching encryption jwk